jdaviz/app.py

```diff
@@ -91,206 +133,567 @@ def __init__(self, configuration=None, *args, **kwargs):
+        return self._viewer_by_reference(reference)
+
+    def get_viewer_data(self, reference, label=None, cls=None):
+        """
+        Returns each data component currently rendered within a viewer
+        instance. Viewers themselves store a default data type to which the
+        Glue data components are transformed upon retrieval. This can be
+        optionally overridden with the ``cls`` keyword.
+
+        Notes
+        -----
+            This is only used in cases where the viewers have been pre-defined
+            in the configuration file. Otherwise, viewers are not stored via
+            reference.
+
+        Parameters
+        ----------
+        reference : str
+            The reference to the viewer defined with the ``reference`` key
+            in the yaml configuration file.
+        label : str, optional
+            Optionally provide a label to retrieve a specific data set from the
+            viewer instance.
+        cls : class
+            The class definition the Glue data components get transformed to
+            when retrieved. This requires that a working set of translation
+            functions exist in the ``glue_astronomy`` package. See
+            ``https://github.com/glue-viz/glue-astronomy`` for more info.
+
+        Returns
+        -------
+        data : list
+            A list of the transformed Glue data objects.
+        """
+        viewer = self.get_viewer(reference)
+        cls = cls or viewer.default_class
+
+        data = [layer_state.layer.get_object(cls=cls)
+                if cls is not None else layer_state.layer
                 for layer_state in viewer.state.layers
                 if hasattr(layer_state, 'layer') and
-                isinstance(layer_state.layer, BaseData)]
+                isinstance(layer_state.layer, BaseData)
+                and (label is None or layer_state.layer.label == label)]
 
-        if cls is not None:
-            data = [d.get_object(cls) for d in data]
+        if label is not None:
+            return next(iter(data), None)
 
         return data
 
-    def add_data(self, reference, data, name=""):
-        name = name or "New Data"
+    def add_data(self, data, name=None):
+        """
+        Add data to the Glue ``DataCollection``.
+
+        Parameters
+        ----------
+        data : any
+            Data to be stored in the ``DataCollection``. This must either be
+            a `~glue.core.data.Data` instance, or an arbitrary data instance
+            for which there exists data translation functions in the
+            glue astronomy repository.
+        name : str, optional
+            The name associated with this data. If none is given, a generic
+            name is generated.
+        """
+        # Include the data in the data collection
         self.data_collection[name or "New Data"] = data
 
-        viewer = self.viewers(reference)
-        viewer.add_data(name)
+    def set_viewer_data(self, reference, label):
```

Comment:
Two "easy" things here:
1. I think the name `set` (and `unset` in the other method) is confusing because it implies "set the viewer to a specific data", but of course you can set *multiple* data (I had to check this because the name confused me).  What about `add`/`remove`?  Or possibly `show`/`hide`?
2. `reference` and `label` are ambiguous variable names (and the user will be seeing this if they look at the function signature).  How about `viewer_reference`, and `data_label`?

Now three possibly "harder" things (i.e., if you want to punt it to a later PR that's fine, but just make sure we have an issue that records it if it doesn't get in here):
1. I think this should do *something* if the user tries to add data that's already been added.  The simplest is to just yield a warning, but that might be annoying.  The alternative (my preference) is to raise an exception, but add a keyword argument to the method (and same to `unset`) that lets the user silence this exception (which defaults to the exception).  Something like a keyword ``allow_repeat_add=False``
2. I think there should be an option (which could either be a keyword here or a separate method) which clears all data *except* the selected one. You already demonstrate the need for that in the `smooth` case, where a user might want to show *just* the smoothed spectrum, not both the smooth and unsmoothed one (there's uses for both cases).  If you take one of my name changes in "easy" 1 above, in fact maybe that's best implemented as `set_viewer_data`, which implies all other viewer data is cleared (and then no need for an `unset`).
3. I think it would be useful to allow a shortcut for the user that combines `add_data` and `set_viewer_data`.  I.e., if `label` does not exist, have the code try to add it (and only raise an exception if that fails).  Then I don't ever have to bother with manually adding data if I want to do something quick and don't care too much about the name

Reply:
> I think this should do something if the user tries to add data that's already been added. The simplest is to just yield a warning, but that might be annoying. The alternative (my preference) is to raise an exception, but add a keyword argument to the method (and same to unset) that lets the user silence this exception (which defaults to the exception). Something like a keyword allow_repeat_add=False

Hmm, nothing bad happens when data is added to a viewer that already contains the data -- should we really even care in this case?

Reply:
> I think there should be an option (which could either be a keyword here or a separate method) which clears all data except the selected one. You already demonstrate the need for that in the smooth case, where a user might want to show just the smoothed spectrum, not both the smooth and unsmoothed one (there's uses for both cases). If you take one of my name changes in "easy" 1 above, in fact maybe that's best implemented as set_viewer_data, which implies all other viewer data is cleared (and then no need for an unset).

I don't think this needs to be a completely separate method, but instead should just be a keyword argument to `add_viewer_data` (changed from `set_viewer_data`); something like `clear_other_data=True`.

Reply:
> Hmm, nothing bad happens when data is added to a viewer that already contains the data -- should we really even care in this case?

I find it slightly annoying that a duplicate entry gets added to the data selection list. Not a showstopper, but it might be nice to not have duplicates show up.

Reply:
> I find it slightly annoying that a duplicate entry gets added to the data selection list. Not a showstopper, but it might be nice to not have duplicates show up.

I'm not sure I follow? Asking to plot a data set in the viewer doesn't affect the list of data set objects, just toggles on their visibility. Are you seeing duplicate entries when you attempt to plot the same data set?

Reply:
If I run the `app.add_data(smoothed_spec, name="Smooth Spectrum")` cell three times, there will be three `Smooth Spectrum` entries in the data selection dropdown list.

Reply:
Ah, yes. Your use case is definitely different. Yeah, glue doesn't store data via label, so it's okay to internally have the same name, however, this may get confusing. We can certainly raise an error, or alternatively, auto-increment some number at the end of the data label.

Reply:
Re-reading, I can see that maybe I mis-understood Eriks point. @eteq can you expand on 

> I think this should do something if the user tries to add data that's already been added. The simplest is to just yield a warning, but that might be annoying. The alternative (my preference) is to raise an exception, but add a keyword argument to the method (and same to unset) that lets the user silence this exception (which defaults to the exception). Something like a keyword allow_repeat_add=False

Is this when the user uses `add_data` or `add_viewer_data`? The former is adding data to the data collection, which does allow data items with the same label.

Reply:
> @eteq can you expand on ... Is this when the user uses add_data or add_viewer_data? 

Ah, right, sorry that was ambiguous.  I was thinking about `add_viewer_data`.  

Now that you say it, I think that *might* also be desirable in `add_data`, but I think in that case definitely a warning is the right answer (since sometimes there are good reasons to add multiple versions of the same data in the collection), and probably is on the glue side anyway?

Reply:
Sounds good. I've captured these two cases in #116.